aws/retry/adaptive.go

```diff
@@ -0,0 +1,173 @@
+package retry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/ratelimit"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+)
+
+const (
+	// DefaultRequestCost is the cost of a single request from the adaptive
+	// rate limited token bucket.
+	DefaultRequestCost uint = 1
+)
+
+// DefaultThrottles provides the set of errors considered throttle errors that
+// are checked by default.
+var DefaultThrottles = []IsErrorThrottle{
+	ThrottleErrorCode{
+		Codes: DefaultThrottleErrorCodes,
+	},
+}
+
+// AdaptiveModeOptions provides the functional options for configuring the
+// adaptive retry mode, and delay behavior.
+type AdaptiveModeOptions struct {
+	FailOnNoAttemptTokens bool
+	RequestCost           uint
+
+	Throttles []IsErrorThrottle
+
+	MaxAttempts int
+
+	MaxBackoff time.Duration
+	Backoff    BackoffDelayer
+
+	Retryables []IsErrorRetryable
+	Timeouts   []IsErrorTimeout
+
+	RetryRateLimiter RateLimiter
+	RetryCost        uint
+	RetryTimeoutCost uint
+	NoRetryIncrement uint
+}
+
+// AdaptiveMode provides an experimental retry strategy that expands on the
+// Standard retry strategy, adding client send rate limits. The send rate limit
+// is initially unrestricted, but becomes restricted when the attempt fails
+// with for a throttle error. Eventually unrestricted send rate limit is
+// restored once attempts no longer are failing for throttle errors.
+type AdaptiveMode struct {
+	options   AdaptiveModeOptions
+	throttles IsErrorThrottles
+
+	retryer   aws.RetryerV2
+	rateLimit *adaptiveRateLimit
+}
+
+// NewAdaptiveMode returns an initialized AdaptiveMode retry strategy.
+func NewAdaptiveMode(optFns ...func(*AdaptiveModeOptions)) *AdaptiveMode {
+	adaptiveOptions := AdaptiveModeOptions{
+		RequestCost: DefaultRequestCost,
+		Throttles:   DefaultThrottles,
```

Comment:
Should this be a copy as to avoid potential for mutations of the package global slice?

Reply:
Will fix that up. The `options.Throttles` will be copied within the constructor, but does leave room for the global var to be duplicated in the options.  Will moves these into some kind of copy.